### ISSUE TITLE
Replace deprecated `who` argument from user query

### ISSUE THREAD

**Issue by grappler** [WordPress/gutenberg#39986]
## What problem does this address?
In WordPress 5.9 the `who` argument has been deprecated. [Trac Ticket](https://core.trac.wordpress.org/ticket/16841)

Gutenberg is still using the `who` argument and recommending it to be used.
https://github.com/WordPress/gutenberg/blob/08cf4a96d1f14fcdf4c55886ba237fb294f9c465/packages/editor/src/components/post-author/constants.js#L2
https://github.com/WordPress/gutenberg/blob/23f11a18726bf69717595451ab3250cb67b426f6/packages/block-library/src/avatar/user-control.js#L11
https://github.com/WordPress/gutenberg/blob/08cf4a96d1f14fcdf4c55886ba237fb294f9c465/packages/block-library/src/query/edit/inspector-controls/author-control.js#L15
https://github.com/WordPress/gutenberg/blob/760815a731805fa05f6cbb234660b254ad1ca6c9/packages/block-library/src/post-author/edit.js#L43
https://github.com/WordPress/gutenberg/blob/711ea0eda4e3b4d28b9e0d70723ec04622dec278/packages/core-data/README.md?plain=1#L26

## What is your proposed solution?
My suggestion would be to use the capability argument. The following code snippet fetches the `edit_post` capability for the current post type. This code is only compatible with WordPress 5.9. Either Gutenberg will need to increase the minimum supported version or include a check and a fallback with `who: 'authors'`.

```js
const AUTHORS_QUERY = {
	per_page: 50,
	_fields: 'id,name',
	context: 'edit', // Allows non-admins to perform requests.
};
const currentPostType = wp.data.select('core/editor').getCurrentPostType();
const editPostCap = wp.data.select('core').getPostType( currentPostType ).capabilities.edit_post;
wp.data.select('core').getUsers( { ...AUTHORS_QUERY, capability: [ editPostCap ] } );
```

So that everyone does not have to figure out the `edit_post` capability, it might be best to create a helper function. Though in WordPress 5.9 the `getAuthors` function was deprecated https://github.com/WordPress/gutenberg/pull/33725

Related Issues: https://github.com/WordPress/gutenberg/issues/39035 & https://github.com/WordPress/gutenberg/issues/17364

**Comment by ocean90** [comment]
Noting that the current request already generates a `X-WP-DeprecatedParam` header for this:

> `X-WP-DeprecatedParam: WP_User_Query (since 5.9.0; <code>who</code> is deprecated. Use <code>capability</code> instead.)`

@TimothyBJacobs @spacedmonkey @youknowriad Do you have some ideas how we can detect whether `who` or `capability` should be used?

**Comment by TimothyBJacobs** [comment]
I'm not sure what the progress is on the client having knowledge of the REST API schemas, cc @youknowriad. But the best way would be to look at the schema for the endpoint and check for the presence of the `capability` property in the `GET` args.

**Comment by grappler** [comment]
The only reason that I can see why we would need to check for the support is to be able to run the Gutenberg plugin on WordPress 5.8. Depending on the time it takes for a patch, WordPress 6.0 will have been released.

It might make more sense to increase the minimum required WP version.
https://github.com/WordPress/gutenberg/blob/956a836ce8fcdd61b186cf1c2d1047b48126898d/gutenberg.php#L6

**Comment by ocean90** [comment]
As of #41306 the plugin now requires 5.9 so I guess we can switch the query to use `capability` now.

**Comment by Mamaduka** [comment]
The user's controller has special permission handling for `'authors' === $request['who']`, which allows editors to render the authors' selection for low-capability users. We need to keep that in mind when making the change.

https://github.com/WordPress/wordpress-develop/blob/3f9317c2aa2ba650c0d930acaca1bb840c95a6da/src/wp-includes/rest-api/endpoints/class-wp-rest-users-controller.php#L236-L251

Somewhat related - https://core.trac.wordpress.org/ticket/56429.

**Comment by apermo** [comment]
I just ran into this issue. I was asked to create a new role for guest authors, so basically a user that will never log in, but shall be selected by the editors to show in the frontend.
I read that granting the role `edit_posts` should be enough for the custom role to appear. After entering the rabbit hole I ended up here.

I'm now at a point where I either will have to add a user lvl of at least 1, in order to satisfy the deprecated filter, or alter the User Query, which might be incompatible with future updates.

TLDR: Sticking with the `who` parameter has implications that you cannot easily add custom roles to the "Select Author Input" in Gutenberg. 

Example code:

```
add_role(
	'guest_author',
	'Guest Author',
	[
		'read' => true,
		'edit_posts' => true,
		'publish_posts' => true,
	]
);
```

Create new User with the given role, try to find it in "Author".

Props to @zaphir for helping me find the issue.
